sql/migrate/migrate.go

```diff
@@ -587,6 +609,37 @@ func (r *Revision) setSQLErr(err error, stmt string) error {
 	return err
 }
 
+type (
+	// A Logger logs migration execution.
+	Logger interface {
+		Log(LogEntry)
+	}
+
+	// LogEntry marks several types of logs to be passed to a Logger.
+	LogEntry interface {
+		logEntry()
+	}
+
+	// LogExecution is sent once when execution of the migration files has been started.
+	// It holds the filenames of the pending migration files.
+	LogExecution []string
+
+	// LogFile is sent if a new migration file is executed.
+	LogFile struct {
+		// Version executed.
+		Version string
+		// Desc of migration executed.
+		Desc string
+	}
+
+	// LogStmt is sent if a new SQL statement is executed.
+	LogStmt string
+)
```

Comment:
Looks good, but I suggest using struct for all different types as it's easy to extend without introducing breaking changes.